metadata-ingestion/src/datahub/testing/mcp_diff.py

```diff
@@ -0,0 +1,268 @@
+            aspects_removed=aspects_removed,
+            aspects_changed=aspects_changed,
+        )
+
+
+@dataclass
+class MCPDiff:
+    aspect_changes: Dict[str, Dict[str, MCPAspectDiff]]  # urn -> aspect -> diff
+    urns_added: Set[str]
+    urns_removed: Set[str]
+
+    def __bool__(self) -> bool:
+        return bool(self.aspect_changes)
+
+    @classmethod
+    def create(
+        cls,
+        golden: AspectsByUrn,
+        output: AspectsByUrn,
+        ignore_paths: Sequence[str],
+    ) -> "MCPDiff":
+        ignore_paths = [cls.convert_path(path) for path in ignore_paths]
+
+        aspect_changes: Dict[str, Dict[str, MCPAspectDiff]] = defaultdict(dict)
+        for urn in golden.keys() | output.keys():
+            golden_map = golden.get(urn, {})
+            output_map = output.get(urn, {})
+            for aspect_name in golden_map.keys() | output_map.keys():
+                diff = DeepDiff(
+                    t1=golden_map.get(aspect_name, []),
+                    t2=output_map.get(aspect_name, []),
+                    exclude_regex_paths=ignore_paths,
+                    ignore_order=True,
+                    custom_operators=[DeltaInfoOperator()],
+                )
+                if diff:
+                    aspect_changes[urn][aspect_name] = MCPAspectDiff.create(diff)
+
+        return cls(
+            urns_added=output.keys() - golden.keys(),
+            urns_removed=golden.keys() - output.keys(),
+            aspect_changes=aspect_changes,
+        )
+
+    @staticmethod
+    def convert_path(path: str) -> str:
+        # Attempt to use paths intended for the root golden... sorry for the regex
+        return re.sub(
+            r"root\\?\[([0-9]+|\\d\+)\\?]\\?\['aspect'\\?](\\?\['(json|value)'\\?])?",
+            r"root\[\\d+].aspect",
+            path,
+        )
+
+    def apply_delta(self, golden: List[Dict[str, Any]]) -> None:
+        aspect_diffs = [v for d in self.aspect_changes.values() for v in d.values()]
+        for aspect_diff in aspect_diffs:
+            for (_, old, new), diffs in aspect_diff.aspects_changed.items():
+                golden[old.delta_info.idx] = new.delta_info.original
+
+        indices_to_remove = set()
+        for aspect_diff in aspect_diffs:
+            for ga in aspect_diff.aspects_removed.values():
+                indices_to_remove.add(ga.delta_info.idx)
+        for idx in sorted(indices_to_remove, reverse=True):
+            del golden[idx]
+
+        for aspect_diff in aspect_diffs:  # Ideally would have smarter way to do this
+            for ga in aspect_diff.aspects_added.values():
+                golden.insert(ga.delta_info.idx, ga.delta_info.original)
+
+    def pretty(self, verbose: bool = False) -> str:
+        """The pretty human-readable string output of the diff between golden and output."""
+        s = []
+        for urn in self.urns_added:
+            s.append(f"Urn added, {urn}{' with aspects:' if verbose else ''}")
```

Comment:
i wonder if we could add color / bolding to make the output even more readable?

might be cool if it's not too hard - `click.style` does some of this stuff, which we use in pipeline.py

Reply:
In interest of time, will leave this as a potential followup